### PR TITLE
WIP Write up some usecases

### PR DESCRIPTION
**Pull request by chmelevskij**
Some basic use cases for formidable. Just added stuff from top of my head and what I was using it for atm. Could expand on some.

### CODE REVIEW

File: docs/usecases.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Adopting Existing Design Systems
+
+In some cases there is existing design schema in the company or there is a need
+to use component library (Material, Ant, Bootstrap etc.). There have to be
+several optins available:
+  1. If only few components are needed, user can import those components and
+     register them with the form instance, either registering them as new or
+     overriding the default.
+  2. In cases where whole replacement is needed form needs to allow passing
```
Comment (wittydeveloper) [inline docs/usecases.md]: Totally agree on this, this is a great idea (totally in the move of having a `kit=` `<Form>` prop 😉

File: docs/usecases.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Usecases
+
+## Batteries Included With access to schema for Web
+
+User of the library has GraphQL endpoint and access to the schema (E.G. Full
+stack developer) with constraint directives. Think of schema from `gql` tag.
+Forms have to be automatically generated, meaning all the user needs to do is
+run schema:generate and then just include generated files where needed. This
+gives form and it's validation out of the box with default styling. Good for
```
Comment (wittydeveloper) [inline docs/usecases.md]: 👍 👍 👍

**Comment by chmelevskij** [comment]
I see the point of skipping the generation all together. One thing which needs some investigation is the possibility to use `@constraint` directive. At the moment, directives are not exposed to the client, so we would need to investigate could we do with client-side directives?

I've already encountered an issue with typescript type validation and client side schema, it doesn't seem to work.

**Comment by wittydeveloper** [comment]
@chmelevskij I took a quick look to `apollo-client` and saw no evidence that we could implement what I propose, so I think that we should stick with a generation from API step in the first version.
We'll iterate for a better Developer Experience later.

More, I do see on the [graphql-js](https://github.com/graphql/graphql-js/) repo that `IntrospectionQuery` do return all directives defined in the `Schema`.

See here: https://github.com/graphql/graphql-js/blob/926e4d80c558b107c49e9403e943086fa9b043a8/src/utilities/introspectionQuery.js#L21-L38

Does this solves the issue of generation?

**Comment by wittydeveloper** [comment]
Asked a question regarding having the Schema on client-side with Apollo to a core developer of Apollo GraphQL on Twitter: https://twitter.com/whereischarly/status/1094642418237014017

**Comment by chmelevskij** [comment]
It seems a bit weird... `IntrospectionQuery` does expose the directives itself, but it doesn't show which fields have them applied. If you grab a look at https://github.com/graphql/graphql-js/blob/926e4d80c558b107c49e9403e943086fa9b043a8/src/utilities/introspectionQuery.js#L40 those are missing there

**Comment by wittydeveloper** [comment]
@chmelevskij Yeah, you're right, I totally missed the point about directives.

I'm gonna take a look on how we could make `@constraint` works.
Do yo have any ideas? 🧠

**Comment by chmelevskij** [comment]
Well `graphql-tag` seems to add the directives [see example](https://codesandbox.io/s/6vr7kx9lwk) if you look in the console you can see those.

I've played with the server side exported typedefs, they seem to have it as well. Only that developer needs to have the access to server side code.

Another option would be to use client side directives when using `gql`. As you documented out in [here](https://github.com/wittydeveloper/react-formidable/blob/de4c2056f79d2c424646160087c4577102027b7d/docs/api/components.md) similar to how `apollo` cli generates types we could do a pass to generate the validations.

**Pull request closed without merging**